rmw_dds_common/src/graph_cache.cpp

```diff
@@ -0,0 +1,1039 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rmw_dds_common/graph_cache.hpp"
+
+#include <algorithm>
+#include <cassert>
+#include <cstring>
+#include <functional>
+#include <iterator>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <ostream>
+#include <set>
+#include <sstream>
+#include <string>
+#include <tuple>
+#include <utility>
+#include <vector>
+
+#include "rcutils/strdup.h"
+
+#include "rmw/convert_rcutils_ret_to_rmw_ret.h"
+#include "rmw/error_handling.h"
+#include "rmw/sanity_checks.h"
+#include "rmw/topic_endpoint_info.h"
+#include "rmw/topic_endpoint_info_array.h"
+
+#include "rmw_dds_common/gid_utils.hpp"
+
+using rmw_dds_common::GraphCache;
+using rmw_dds_common::operator<<;
+
+static const char log_tag[] = "rmw_dds_common";
+
+#define GRAPH_CACHE_CALL_ON_CHANGE_CALLBACK_IF(condition) \
```

Comment:
@ivanpauno nit^N: I find 
```suggestion
#define GRAPH_CACHE__CALL_ON_CHANGE_CALLBACK__IF(condition) \
```
easier to grok.

Reply:
Another nit^N: require a force cache, i.e. by passing `this`, so that it reads better when used

Reply:
> @ivanpauno nit^N: I find
> 
> easier to grok.

Double underscores are reserved to be used by the standard library.
I remember somebody complaint about that we use double underscores in include guards, so I wouldn't extend that pattern.

> Another nit^N: require a force cache, i.e. by passing this, so that it reads better when used

What do you mean?
Do you mean changing the macro to:
```cpp
#define GRAPH_CACHE_CALL_ON_CHANGE_CALLBACK_IF(graph_cache, condition) \
```

Reply:
> Double underscores are reserved to be used by the standard library.

I see, fair enough.

> Do you mean changing the macro to:

Correct :)

Reply:
Not major, but this one remains unaddressed.

Reply:
I forgot about that one, addressed in https://github.com/ros2/rmw_dds_common/pull/4/commits/7579823e226e55aaac812b2fa9182dbe1fec5472.